### ISSUE TITLE
Client never asks for work when there are stalled downloads

### ISSUE THREAD

**Issue by Rytiss** [BOINC/boinc#3193]
**Describe the bug**
I created some work using a temporary download server. The download server was then terminated, but some hosts were still downloading. They are now unable to download, and they no longer ask for work: "Not requesting tasks: some download is stalled".

The downloads never terminate as they point to a non-existent DNS name ("can't resolve hostname") - download termination is done when a 404 error exists, but not when DNS name does not exist.

I cannot obtain the DNS name again as the provider does not allow one to choose names and instead generates random ones.

**Steps To Reproduce**
1. Create work with data files stored on a temporary domain name (e.g. AWS S3 bucket)
2. Download work, but terminate client while the downloads are still in progress.
3. Terminate the temporary download server
4. Restart the client. Downloads will now fail, and client won't ever ask for more work.

**Expected behavior**
Terminate downloads after some time of continuing failures (e.g. 72 hours).

**Comment by RichardHaselgrove** [comment]
The 'stalled download' block on requesting new tasks only applies to the project with the download problem: work for other projects should be fetched normally.

So this is untidy, but not so much of a problem as the wording may make it sound.

**Comment by Rytiss** [comment]
It is a huge problem for the project, as it can essentially drop all of its hosts (this is actually what has happened in my case).

**Comment by davidpanderson** [comment]
There is a time-out on file transfers; currently it's 90 days.
How about if we change this to 5 days?

**Comment by RichardHaselgrove** [comment]
No, please don't. The original timeout was 14 days, and it was increased because CPDN had some valuable data out in the field when one of their upload servers failed. It took longer than 14 days to source, specify, fund, order, take delivery, install, configure, and deploy a new server.

I think the takeaway message is 'never use a temporary url on a live project'. But since it's happened:

Can the users who are paying attention abort the transfers from BOINC Manager?

Can the users who aren't paying attention be saved by cancelling the jobs (I assume the project scheduler is on a static url, so they can still contact it). @davidpanderson - would a server abort cancel the client downloads? If not, should it? Possibly depends whether the original problem affected task data files, or application executables.

**Comment by davidpanderson** [comment]
BOINC currently has no mechanism by which a project can abort/cancel jobs on the client.  Maybe it should.

**Comment by RichardHaselgrove** [comment]
Are you sure? I'm sure I've seen jobs cancelled remotely over the years. Why do we have

`#define EXIT_ABORTED_BY_PROJECT     202`

**Comment by Rytiss** [comment]
Tasks get aborted, downloads are not.

**Comment by RichardHaselgrove** [comment]
I'll try and trace back that trac # 366 fix (edited into previous post) over the weekend. No time this morning - I've got a hospital appointment.

**Comment by RichardHaselgrove** [comment]
The updated references for this problem (historical version) in github are issue #366 and commit 4a8b4fae2c7ab07baad0cda490fea60ee2d0542d

David's checkin note 'may fix' sounds uncertain - perhaps @Rytiss could investigate. Does garbage-collect run continuously, or only at client restart?

**Comment by Rytiss** [comment]
My file is marked as sticky:

```
<file>
    <name>jf_2ac70f61f5c32583afc27acdb3aa3643</name>
    <nbytes>772953.000000</nbytes>
    <max_nbytes>0.000000</max_nbytes>
    <md5_cksum>2ac70f61f5c32583afc27acdb3aa3643</md5_cksum>
    <status>0</status>
    <sticky/>
    <download_url>(redacted)</download_url>
    <persistent_file_xfer>
        <num_retries>132</num_retries>
        <first_request_time>1555611615.885090</first_request_time>
        <next_request_time>1561809487.540644</next_request_time>
        <time_so_far>299.687647</time_so_far>
        <last_bytes_xferred>0.000000</last_bytes_xferred>
        <is_upload>0</is_upload>
    </persistent_file_xfer>
</file>
```

Because it is sticky, and status=0, it is considered to have ref_cnt >= 1 (https://github.com/BOINC/boinc/blob/master/client/client_state.cpp#L1738). Since it is considered referenced, persistent file transfer is not deleted: https://github.com/BOINC/boinc/blob/master/client/client_state.cpp#L1748

I do not see the 90 day timeout in the code, but I might be missing something. If that is present, I will know in 19 days (it is currently 71 days into trying to download).

**Comment by RichardHaselgrove** [comment]
The change to 90 days was made in commit 7bca6c021badf196fd67f15f5ebb7e758db6c4e8 (amongst other things - see final file reference). As I said, the context was result uploads, but at the time (BOINC v6.6), I don't think upload and download transfers were distinguished.

**Comment by Rytiss** [comment]
Thanks, that's helpful - the downloads should indeed time out in 90 days: https://github.com/BOINC/boinc/blob/f14d96d2be2e5a290177f7b5391b0802fbd3756c/client/pers_file_xfer.cpp#L340

This does not have a distinction between downloads and uploads. I do recognize Richard's case of project servers not being able to accept uploads; however, downloads should be totally under project's control, and workunits would just get resent to other hosts if not terminated from the server side, so I suggest the following:

* terminate uploads as is, after 90 days;
* terminate downloads after 5 days;
* do not block work requests when there are stalled uploads (this is not needed for my problem, but could occur elsewhere, so it is best to just fix that).

**Comment by RichardHaselgrove** [comment]
I agree with your first two points: the grace period to complete downloads could be even shorter, because we know the user had a working internet connection when work was requested and allocated. If the project can't keep all their servers working at the same time, they probably shouldn't be allocating work to volunteers in the first place.

I think the third point is more problematic. That would allow an unbounded build-up of completed work to occupy disk space on the volunteer's hard drive in the case of extended failures, and massive congestion on the upload pipeline once service was restored. On the other hand, the current cut-off at twice the number of CPU cores is reached very quickly on fast GPU machines: volunteers who pay attention get quite frustrated by that. We should think more broadly about the needs of projects at both ends of the spectrum before making changes.

**Comment by Rytiss** [comment]
The project is also in control for the problem you describe. If they know they will have resources for uploads, they can continue issuing work; otherwise they can stop and wait for enough uploads to happen. There's already intervention needed within the project in case of outages, e.g. stop resending work, extend deadlines on the server side, etc.

**Comment by Aurum420** [comment]
Currently the "Not requesting tasks: some download is stalled" error is appearing randomly while attempting to DL only ARP WUs from WCG. Once the error occurs it also blocks DLing OPN and MCM CPU WUs. Rebooting does not clear it. It seems to clear itself after some random number of hours but in less than a day. If one selects "Allow research to run on my CPU? = No" in the WCG device profile then OPNG GPU WUs will DL immediately.
I don't understand the discussion above but it does not jive with what I observe about the behavior of this error. It seems it's a  project-specific server-side error.
